integration/csv_fork_test.go

```diff
@@ -0,0 +1,687 @@
+func TestBIP0113Activation(t *testing.T) {
+	t.Parallel()
+
+	// Initialize the primary mining node with only the genesis block.
+	btcdCfg := []string{"--rejectnonstd"}
+	r, err := rpctest.New(&chaincfg.SimNetParams, nil, btcdCfg)
+	if err != nil {
+		t.Fatal("unable to create primary harness: ", err)
+	}
+	if err := r.SetUp(true, 10); err != nil {
+		t.Fatalf("unable to setup test chain: %v", err)
+	}
+	defer r.TearDown()
+
+	// Create a fresh output for usage within the test below.
+	const outputValue = btcutil.SatoshiPerBitcoin
+	outputKey, testOutput, testPkScript, err := makeTestOutput(r, t,
+		outputValue)
+	if err != nil {
+		t.Fatalf("unable to create test output: %v", err)
+	}
+
+	// Fetch a fresh address from the harness, we'll use this address to
+	// send funds back into the Harness.
+	addr, err := r.NewAddress()
+	if err != nil {
+		t.Fatalf("unable to generate address: %v", err)
+	}
+	addrScript, err := txscript.PayToAddrScript(addr)
+	if err != nil {
+		t.Fatalf("unable to generate addr script: %v", err)
+	}
+
+	// Now create a transaction with a lock time which is "final" according
+	// to the latest block, but not according to the current median time
+	// past.
+	tx := wire.NewMsgTx(1)
+	tx.AddTxIn(&wire.TxIn{
+		PreviousOutPoint: *testOutput,
+	})
+	tx.AddTxOut(&wire.TxOut{
+		PkScript: addrScript,
+		Value:    outputValue - 1000,
+	})
+
+	// We set the lock-time of the transaction to just one minute after the
+	// current MTP of the chain.
+	chainInfo, err := r.Node.GetBlockChainInfo()
+	if err != nil {
+		t.Fatalf("unable to query for chain info: %v", err)
+	}
+	tx.LockTime = uint32(chainInfo.MedianTime) + 1
+
+	sigScript, err := txscript.SignatureScript(tx, 0, testPkScript,
+		txscript.SigHashAll, outputKey, true)
+	if err != nil {
+		t.Fatalf("unable to generate sig: %v", err)
+	}
+	tx.TxIn[0].SignatureScript = sigScript
+
+	// This transaction should be rejected from the mempool as using MTP
+	// for transactions finality is now a policy rule. Additionally, the
+	// exact error should be the rejection of a non-final transaction.
+	_, err = r.Node.SendRawTransaction(tx, true)
+	if err == nil {
+		t.Fatalf("transaction accepted, but should be non-final")
+	} else if !strings.Contains(err.Error(), "not finalized") {
+		t.Fatalf("transaction should be rejected due to being "+
+			"non-final, instead :%v", err)
+	}
+
+	// However, since the block validation consensus rules haven't yet
+	// activated, a block including the transaction should be accepted.
+	txns := []*btcutil.Tx{btcutil.NewTx(tx)}
+	block, err := r.GenerateAndSubmitBlock(txns, -1, time.Time{})
+	if err != nil {
+		t.Fatalf("unable to submit block: %v", err)
+	}
+	txid := tx.TxHash()
+	assertTxInBlock(r, t, block.Hash(), &txid)
+
+	// Next, mine enough blocks to ensure that the soft-fork becomes
+	// activated. Assert that the block version of the second-to-last block
+	// in the final range is active.
+	numBlocks := (r.ActiveNet.MinerConfirmationWindow * 2) + 1
```

Comment:
Hmm, yes I think you're correct, CSV should be active at 300, but _not_ at height 299. 

As we discussed offline, I'll separate out the `getblockchaininfo` portion of this PR into a new one, so we can use it to write some additional integration tests for bip9 so we can drill down into the issue an locate where the off-by-one error is.